### PR TITLE
Add deflake handler

### PR DESCRIPTION
**Pull request by chunhtai**
Moved some methods from the file_flaky_issue_and_pr handler to the utils. Added the new handler with a bunch of tests

### CODE REVIEW

File: app_dart/lib/src/request_handlers/deflake_flaky_builders.dart [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    final YamlMap ci = loadYaml(content) as YamlMap;
+    final YamlList targets = ci[kCiYamlTargetsKey] as YamlList;
+    final List<YamlMap> flakyTargets = targets
+        .where((dynamic target) => target[kCiYamlTargetIsFlakyKey] == true)
+        .map<YamlMap>((dynamic target) => target as YamlMap)
+        .toList();
+    final List<_BuilderInfo> result = <_BuilderInfo>[];
+    final List<String> lines = content.split('\n');
+    final Map<String, PullRequest> nameToExistingPRs = await getExistingPRs(gitHub, slug);
+    for (final YamlMap flakyTarget in flakyTargets) {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/deflake_flaky_builders.dart:61]: As we talked about, please add a list to exclude builders that we want to keep them as flaky all the time, like the ios32 ones.

File: app_dart/lib/src/request_handlers/deflake_flaky_builders.dart [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    final BigqueryService bigquery = await config.createBigQueryService();
+    final String ciContent = await gitHub.getFileContent(slug, kCiYamlPath);
+    final List<_BuilderInfo> eligibleBuilders = await _getEligibleFlakyBuilders(gitHub, slug, content: ciContent);
+    String testOwnerContent;
+    for (final _BuilderInfo info in eligibleBuilders) {
+      final List<BuilderRecord> builderRecords =
+          await bigquery.listRecentBuildRecordsForBuilder(kBigQueryProjectId, builder: info.name, limit: kRecordAmount);
+      if (builderRecords.every((BuilderRecord record) => !record.isFlaky)) {
+        testOwnerContent ??= await gitHub.getFileContent(slug, kTestOwnerPath);
+        await _fileDeflakePullRequest(gitHub, slug,
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/deflake_flaky_builders.dart:41]: Nit: `_fileDeflakePullRequest` -> `_deflakyPullRequest`.

File: app_dart/lib/src/request_handlers/deflake_flaky_builders.dart [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        lines.removeAt(nextLine);
+        return lines.join('\n');
+      }
+      nextLine += 1;
+    }
+    throw 'Cannot find the flaky flag, is the test really marked flaky?';
+  }
+}
+
+class _BuilderInfo {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/deflake_flaky_builders.dart:135]: Nit: Doc

File: app_dart/lib/src/request_handlers/flaky_handler_utils.dart [hunk 4]
```diff
@@ -119,7 +137,34 @@ class PullRequestBuilder {
...
   String get pullRequestTitle => 'Marks ${statistic.name} to be flaky';
-  String get pullRequestBody => '${_buildHiddenMetaTags(statistic)}Issue link: ${issue.htmlUrl}\n';
+  String get pullRequestBody => '${_buildHiddenMetaTags(name: statistic.name)}Issue link: ${issue.htmlUrl}\n';
+}
+
+/// A builder to build the pull request title and body for marking test flaky
+class DeflakePullRequestBuilder {
+  DeflakePullRequestBuilder({
+    @required this.name,
+    @required this.recordsAmount,
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/flaky_handler_utils.dart:147]: Same here, nit: recordsNumber.

File: app_dart/lib/src/request_handlers/flaky_handler_utils.dart [hunk 5]
```diff
@@ -119,7 +137,34 @@ class PullRequestBuilder {
...
+    this.issue,
+  });
+
+  final String name;
+  final Issue issue;
+  final int recordsAmount;
+
+  String get pullRequestTitle => 'Marks $name to be not flaky';
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/flaky_handler_utils.dart:155]: Nit: not flaky -> unflaky?

File: app_dart/lib/src/request_handlers/flaky_handler_utils.dart [hunk 6]
```diff
@@ -119,7 +137,34 @@ class PullRequestBuilder {
...
+  String get pullRequestBody {
+    String body = _buildHiddenMetaTags(name: name);
+    if (issue != null) {
+      body +=
+          'The issue ${issue.htmlUrl} has been closed, and the test has been passing for [$recordsAmount consecutive runs](${Uri.encodeFull('$_flakeRecordPrefix"$name"')}).\n';
+    } else {
+      body +=
+          'The test has been passing for [$recordsAmount consecutive runs](${Uri.encodeFull('$_flakeRecordPrefix"$name"')}).\n';
+    }
+    body += 'This test can be marked as not flaky.\n';
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/flaky_handler_utils.dart:165]: unflaky.

File: app_dart/lib/src/request_handlers/deflake_flaky_builders.dart [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+import '../request_handling/api_request_handler.dart';
+import '../request_handling/authentication.dart';
+import '../request_handling/body.dart';
+import '../service/bigquery.dart';
+import '../service/config.dart';
+import '../service/github_service.dart';
+import 'flaky_handler_utils.dart';
+
+@immutable
+class DeflakeFlakyBuilders extends ApiRequestHandler<Body> {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/deflake_flaky_builders.dart:20]: Can we add docs for classes and functions? Also for earlier added two APIs? I believe they would be helpful for whoever will work on these APIs later.

File: app_dart/test/request_handlers/deflake_flaky_builders_test.dart [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+      );
+      tester = ApiRequestHandlerTester(request: request);
+
+      handler = DeflakeFlakyBuilders(
+        config,
+        auth,
+      );
+    });
+
+    test('Can create pr if the flaky test is no longer flaky', () async {
```
Comment (keyonghan) [inline app_dart/test/request_handlers/deflake_flaky_builders_test.dart:106]: Nit: with a closed issue.

File: app_dart/lib/src/request_handlers/deflake_flaky_builders.dart (hunk 1 above)
Comment (chunhtai) [reply app_dart/lib/src/request_handlers/deflake_flaky_builders.dart:61]: Do you have a list?

Comment (keyonghan) [reply app_dart/lib/src/request_handlers/deflake_flaky_builders.dart:61]: At least these two for now: `Mac_ios flutter_gallery__transition_perf_e2e_ios32` and `Mac_ios native_ui_tests_ios32`.

File: app_dart/lib/src/request_handlers/deflake_flaky_builders.dart (hunk 7 above)
Comment (chunhtai) [reply app_dart/lib/src/request_handlers/deflake_flaky_builders.dart:32]: added

File: app_dart/test/request_handlers/deflake_flaky_builders_test.dart [hunk 9]
```diff
@@ -200,7 +200,7 @@ void main() {
       expect(result['Status'], 'success');
     });
 
-    test('Can create pr if the flaky test is no longer flaky - No Issue attached', () async {
+    test('Can create pr if the flaky test is no longer flaky without a issue', () async {
```
Comment (keyonghan) [inline app_dart/test/request_handlers/deflake_flaky_builders_test.dart:203]: Nit: an

File: app_dart/lib/src/request_handlers/deflake_flaky_builders.dart [hunk 10]
```diff
@@ -35,18 +54,22 @@ class DeflakeFlakyBuilders extends ApiRequestHandler<Body> {
...
-            info: info, ciContent: ciContent, testOwnerContent: testOwnerContent);
+        await _deflakyPullRequest(gitHub, slug, info: info, ciContent: ciContent, testOwnerContent: testOwnerContent);
       }
     }
     return Body.forJson(const <String, dynamic>{
       'Status': 'success',
     });
   }
 
+  /// Gets the builders that matches conditions:
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/deflake_flaky_builders.dart:68]: Nit: match

**Review by keyonghan — APPROVED** [review]
LGTM with tiny nits.

**Pull request merged**